File: Crud.Server/Models/Transaction.cs
﻿using System.ComponentModel.DataAnnotations;

namespace Crud.Server.Models
{
    public class Transaction
    {
        [Key]
        public int TransactionId { get; set; }
        public DateTime Date { get; set; }
        public decimal Amount { get; set; }
        public string? Description { get; set; }
        public int? CategoryId { get; set; }
        public string? TransactionType { get; set; } // Receita ou Despesa
        public string? PaymentMethod { get; set; }
        public string? ReceiptImage { get; set; }
        public int? TotalInstallments { get; set; } // Total de parcelas, se aplicável
        public int? InstallmentsPaid { get; set; }  // Parcelas já pagas, se aplicável
        public decimal? InstallmentAmount { get; set; } // Valor de cada parcela, se aplicável
        public DateTime? EndDate { get; set; } // Data de término, se aplicável
        public ExpenseType ExpenseType { get; set; } // Tipo de despesa: Fixa, Avulsa, Parcelada
        public virtual Category? Category { get; set; }
        public int InstallmentsRemaining => TotalInstallments.HasValue ? TotalInstallments.Value - InstallmentsPaid.Value : 0;

        public DateTime? CalculateEndDate()
        {
            if (Date != null && TotalInstallments.HasValue && InstallmentsPaid.HasValue)
            {
                return Date.AddMonths(TotalInstallments.Value - InstallmentsPaid.Value);
            }
            return null;
        }
    }

    public enum ExpenseType
    {
        Fixed = 0,
        Occasional = 1,
        Installment = 2
    }
}

File: crud.client/src/app/models/transaction.ts
export interface Transaction {
  transactionId: number;
  date: Date;  // Ou Date se você estiver usando objetos Date
  amount: number;
  description: string;
  categoryId: number;
  transactionType: string;
  paymentMethod: string;
  receiptImage: string;
  totalInstallments: number;
  installmentsPaid: number;
  installmentAmount: number;
  endDate: Date;  // Ou Date
  expenseType?: number;
  category?: {  // O tipo da categoria pode ser opcional caso não esteja sempre presente
    categoryId: number;
    name: string;
    description: string;
  };
}
